sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java

```diff
@@ -0,0 +1,652 @@
+            // empty
+        }
+
+        public StreamingParameters(int displayDensity, int videoWidth, int videoHeight, int frameRate, int bitrate, int interval) {
+            this.displayDensity = displayDensity;
+            this.videoWidth = videoWidth;
+            this.videoHeight = videoHeight;
+            this.frameRate = frameRate;
+            this.bitrate = bitrate;
+            this.interval = interval;
+        }
+
+        /**
+         * Set displayDensity to a value from DisplayMetrics
+         * @param displayDensity
+         */
+        public void setDisplayDensity(int displayDensity) {
+            this.displayDensity = displayDensity;
+        }
+
+        public int getDisplayDensity() {
+            return displayDensity;
+        }
+
+        public void setVideoWidth(int videoWidth) {
+            this.videoWidth = videoWidth;
+        }
+
+        public int getVideoWidth() {
+            return videoWidth;
+        }
+
+        public void setVideoHeight(int videoHeight) {
+            this.videoHeight = videoHeight;
+        }
+
+        public int getVideoHeight() {
+            return videoHeight;
+        }
+
+        public void setFrameRate(int frameRate) {
+            this.frameRate = frameRate;
+        }
+
+        public int getFrameRate() {
+            return frameRate;
+        }
+
+        public void setBitrate(int bitrate) {
+            this.bitrate = bitrate;
+        }
+
+        public int getBitrate() {
+            return bitrate;
+        }
+
+        public void setInterval(int interval) {
+            this.interval = interval;
+        }
+
+        public int getInterval() {
+            return interval;
+        }
+    }
+
+    /**
+     * Initialization method for VirtualDisplayEncoder object. MUST be called before start() or shutdown()
+     * Will overwrite previously set videoWeight and videoHeight
+     * @param context
+     * @param videoStream
+     * @param presentationClass
+     * @param screenParams
+     * @throws Exception
+     */
+    public void init(Context context, OutputStream videoStream, Class<? extends SdlPresentation> presentationClass, ScreenParams screenParams) throws Exception {
+        if (android.os.Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
+            Log.e(TAG, "API level of 21 required for VirtualDisplayEncoder");
+            throw new Exception("API level of 21 required");
+        }
+
+        mDisplayManager = (DisplayManager)context.getSystemService(Context.DISPLAY_SERVICE);
+
+        mContext = context;
+
+        if(screenParams.getImageResolution().getResolutionHeight() != null){
+            streamingParams.videoHeight = screenParams.getImageResolution().getResolutionHeight();
+        }
+
+        if(screenParams.getImageResolution().getResolutionWidth() != null){
+            streamingParams.videoWidth = screenParams.getImageResolution().getResolutionWidth();
+        }
+
+        sdlOutStream = videoStream;
```

Comment:
Eliminating the usage of a PipedOutputStream for the enhanced video streaming API's will provide a good performance boost.  Since we have direct access to the encoded data, we can ship the data off to the protocol and transport through the video stream writer thread and save processing time.  This requires us to add some user facing accessory methods and interfaces.  Please review the changes in the following commit for more details:
https://github.com/smartdevicelink/sdl_android/commit/b7edfd36ad19855d2205e2e27c4f1a7c4c948cc4